Distinguish always/trivially compactible allocations in libpas/bmalloc

ddegazio opened the pull request:
#### 2d71c4b7c91e0c0bde284c99ada0ae797f2ed15c
<pre>
Distinguish always/trivially compactible allocations in libpas/bmalloc
<a href="https://bugs.webkit.org/show_bug.cgi?id=280294">https://bugs.webkit.org/show_bug.cgi?id=280294</a>
<a href="https://rdar.apple.com/136616671">rdar://136616671</a>

Reviewed by Keith Miller.

Creates a new pas_always_compact_allocation_mode, used for allocations
from memory spans that are known to be contiguous and specially managed
and for which pointer compression is particularly simple. We adopt this
mode for the libpas bootstrap heap and JIT reservation.

* Source/bmalloc/bmalloc/CompactAllocationMode.h:
(bmalloc::asPasAllocationMode):
* Source/bmalloc/bmalloc/IsoHeap.cpp:
(bmalloc::api::isoAllocateCompact):
(bmalloc::api::isoTryAllocateCompact):
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
(bmalloc::api::tzoneAllocateCompact):
(bmalloc::api::tzoneTryAllocateCompact):
* Source/bmalloc/libpas/src/libpas/jit_heap.c:
(jit_heap_try_allocate):
* Source/bmalloc/libpas/src/libpas/pas_allocation_mode.h:
(pas_allocation_mode_get_string):
* Source/bmalloc/libpas/src/libpas/pas_utility_heap.c:

Canonical link: <a href="https://commits.webkit.org/284395@main">https://commits.webkit.org/284395@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c80c01623b541960c4a2720341b1f909f10a874

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48246 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21514 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72923 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19998 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56043 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19894 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54863 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13304 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71918 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44077 "Found 1 new test failure: fast/events/ios/contenteditable-autocapitalize.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59450 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35331 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16877 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18360 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61969 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62691 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17224 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74618 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68099 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12826 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16510 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62356 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/new-content-preserve-3d-ancestor.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12865 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59532 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62398 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15373 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10356 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3968 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89878 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44044 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15925 "Found 2 new JSC binary failures: testapi, testb3, Found 4812 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_splice_515632.js.default, ChakraCore.yaml/ChakraCore/test/Basics/ArrayConcat.js.default, ChakraCore.yaml/ChakraCore/test/Error/variousErrors.js.default, ChakraCore.yaml/ChakraCore/test/Function/prototype.js.default, ChakraCore.yaml/ChakraCore/test/Miscellaneous/HasOnlyWritableDataPropertiesCache.js.default, ChakraCore.yaml/ChakraCore/test/Operators/instanceof.js.default, ChakraCore.yaml/ChakraCore/test/RWC/OneNote.ribbon.js.default, ChakraCore.yaml/ChakraCore/test/Regex/BolEol.js.default, ChakraCore.yaml/ChakraCore/test/Strings/HTMLHelpers.js.default, ChakraCore.yaml/ChakraCore/test/Strings/property_and_index_of_string.js.default ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45121 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46316 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44860 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_allocation_mode.h:
```diff
@@ -31,13 +31,19 @@
+
+    /* We are allocating an object from memory where all addresses within
+       that memory are expected to be trivially compactible, like in the
+       immortal heap. */
```
Is it a requirement that all addresses must be trivially compactible? As the comment is written, it sounds like the caller may need to handle the case where an address isn't compactible, in which case it's hard to understand the difference between this and the "maybe" case from the comments.

ddegazio replied:
Yeah I agree, I'll drop the "expected to be" so it's clear all memory in the range _is_ trivially compactible, without exceptions.

kmiller68 commented on Source/bmalloc/bmalloc/CompactAllocationMode.h:
```diff
@@ -35,8 +35,8 @@
-    Compact,
-    NonCompact
+    NonCompact,
+    Compact
```
Should this be MaybeCompact then? I guess from the caller's perspective they just want a compact allocation but whether they end up getting one is up in the air?

ddegazio replied:
This is intentional. A caller to bmalloc should only care that an allocation is `Compact`, not exactly which mechanism is used to ensure that fact. The "always" v.s. "maybe" distinction is a trait of different heap reservations, which are managed by libpas. So it's bmalloc's job to translate from the higher-level request "give me an allocation with an address I can compress" into the lower-level "allocate memory using libpas, in a heap I know to be always-compactible". bmalloc does currently allocate using `pas_maybe_compact_allocation_mode` only, but if for whatever reason we wanted some future bmalloc allocations to go in a special slab or something that is trivially compactible, it'd be bmalloc's responsibility to know how to do that, not bmalloc's callers passing in a `CompactAllocationMode`. So we limit it to just these two cases at the bmalloc level.

kmiller68 reviewed (approved):
r=me with question.

ddegazio committed:
Distinguish always/trivially compactible allocations in libpas/bmalloc
https://bugs.webkit.org/show_bug.cgi?id=280294
rdar://136616671

Reviewed by Keith Miller.

Creates a new pas_always_compact_allocation_mode, used for allocations
from memory spans that are known to be contiguous and specially managed
and for which pointer compression is particularly simple. We adopt this
mode for the libpas bootstrap heap and JIT reservation.

* Source/bmalloc/bmalloc/CompactAllocationMode.h:
(bmalloc::asPasAllocationMode):
* Source/bmalloc/bmalloc/IsoHeap.cpp:
(bmalloc::api::isoAllocateCompact):
(bmalloc::api::isoTryAllocateCompact):
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
(bmalloc::api::tzoneAllocateCompact):
(bmalloc::api::tzoneTryAllocateCompact):
* Source/bmalloc/libpas/src/libpas/jit_heap.c:
(jit_heap_try_allocate):
* Source/bmalloc/libpas/src/libpas/pas_allocation_mode.h:
(pas_allocation_mode_get_string):
* Source/bmalloc/libpas/src/libpas/pas_utility_heap.c:

Canonical link: https://commits.webkit.org/284395@main
